File: 0509_fibonacci_number/FibonacciNumber.java

import java.util.HashMap;

class FibonacciNumber {

    private final HashMap<Integer, Integer> map = new HashMap<>();

    public int fib(int n) {
        if (n < 2) {
            return n;
        }

        if (map.containsKey(n)) {
            return map.get(n);
        }

        int result = fib(n - 1) + fib(n - 2);
        this.map.put(n, result);
        return result;
    }

    public static void main(String[] args) {
        FibonacciNumber ob = new FibonacciNumber();

        System.out.println("Answer = " + ob.fib(6));
    }
}

File: 0509_fibonacci_number/FibonacciNumber.php
<?php

class FibonacciNumber
{

    /**
     * @param Integer $n
     * @return Integer
     */

    private $map = [];

    function fib($n)
    {
        if ($n < 2) {
            return $n;
        }

        if (isset($this->map[$n])) {
            return $this->map[$n];
        }

        $result = $this->fib($n - 1) + $this->fib($n - 2);
        $this->map[$n] = $result;
        return $result;
    }
}

$ob = new FibonacciNumber();

$answer = $ob->fib(6);

echo "Asnwer is " . $answer . "\n";
